File: julia/Euler/src/bin/p0049.jl

# project euler: problem 49

module Prob0049

import Primes: primes
import ..Util: undigits

function get_prime_tbl(n_digits)
    p_tbl = Dict{Int, Vector{Int}}()

    for p in primes(10 ^ (n_digits - 1), 10 ^ n_digits)
        key = undigits(sort(digits(p)))
        p_tbl[key] = push!(get(() -> Vector{Int}(), p_tbl, key), p)
    end
    p_tbl
end

function solve_0049(n_digits = 4)
    p_tbl = get_prime_tbl(n_digits)
    for lst in values(p_tbl)
        len = length(lst)
        if len < 3
            continue
        end
        sort!(lst)
        for i = 1:(len-2), j = (i+1):(len-1)
            tmp = lst[j] * 2 - lst[i]
            if tmp ∈ lst && lst[i] != 1487
                return foldl((acc, x) -> acc * (10 ^ n_digits) + x, [lst[i], lst[j], tmp]; init = 0)
            end
        end
    end

    @assert false "not reached"
end

end #module

using .Prob0049: solve_0049
export solve_0049

File: clojure/src/project_euler/sol/p0049.clj
(ns project-euler.sol.p0049
  (:require
   [project-euler.lib.math :as math]
   [project-euler.lib.math.prime :as prime]
   [project-euler.lib.util :as util]))

(defn- get-prime-tbl
  [ndigits]
  (loop [primes (reverse (prime/primes (math/pow 10 (dec ndigits)) (math/pow 10 ndigits)))
         tbl {}]
    (if-let [p (first primes)]
      (let [k (util/undigits (sort (util/digits p)))]
        (recur (rest primes) (assoc tbl k (cons p (get tbl k '())))))
      tbl)))

(defn- valid-set?
  [[x y z]]
  (and (= (- y x) (- z y)) (not= x 1487) (not= z 8147)))

(defn solve
  ([]
   (let [ndigits 4
         xf (comp (filter #(>= (count (second %)) 3))
                  (mapcat #(util/combination 3 (second %)))
                  (filter valid-set?))
         ;; The problem statement mentions there exists exactly only one answer.
         [x y z] (first (sequence xf (seq (get-prime-tbl ndigits))))]
     (+ (* x (math/pow 10 (* ndigits 2))) (* y (math/pow 10 ndigits)) z))))
